File: app/src/net/reichholf/dreamdroid/activities/ShareActivity.java
/* © 2010 Stephan Reichholf <stephan at reichholf dot net>
 *
 * Licensed under the Create-Commons Attribution-Noncommercial-Share Alike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 */

package net.reichholf.dreamdroid.activities;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.format.DateFormat;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.compose.ui.platform.ComposeView;

import net.reichholf.dreamdroid.DreamDroid;
import net.reichholf.dreamdroid.Profile;
import net.reichholf.dreamdroid.R;
import net.reichholf.dreamdroid.asynctask.SimpleResultTask;
import net.reichholf.dreamdroid.helpers.ExtendedHashMap;
import net.reichholf.dreamdroid.helpers.NameValuePair;
import net.reichholf.dreamdroid.helpers.SimpleHttpClient;
import net.reichholf.dreamdroid.helpers.enigma2.URIStore;
import net.reichholf.dreamdroid.helpers.enigma2.requesthandler.SimpleResultRequestHandler;
import net.reichholf.dreamdroid.room.AppDatabase;
import net.reichholf.dreamdroid.ui.profiles.ProfileListItem;
import net.reichholf.dreamdroid.ui.share.ShareProfilesListState;
import net.reichholf.dreamdroid.ui.share.ShareProfilesListStateKt;

import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Share / view intent → pick a profile (Compose) → play on the box via MEDIA_PLAYER_PLAY.
 */
public class ShareActivity extends AppCompatActivity implements SimpleResultTask.SimpleResultTaskHandler {
	@NonNull
	public static String LOG_TAG = ShareActivity.class.getSimpleName();

	private SimpleResultTask mSimpleResultTask;
	private SimpleHttpClient mShc;
	private ShareProfilesListState mListState;
	@Nullable
	private ProgressDialog mProgress;
	private String mTitle;

	private List<Profile> mProfiles;
	private final Map<Integer, Profile> mProfilesById = new HashMap<>();

	@Override
	public void onCreate(Bundle savedInstanceState) {
		DreamDroid.setTheme(this);
		super.onCreate(savedInstanceState);
		setContentView(R.layout.share_list_content);
		setTitle(getText(R.string.watch_on_dream));
		Toolbar toolbar = findViewById(R.id.toolbar);
		setSupportActionBar(toolbar);
		mListState = new ShareProfilesListState();
		ComposeView compose = findViewById(R.id.compose_profiles);
		ShareProfilesListStateKt.bindShareProfilesScreen(
				compose,
				mListState,
				item -> {
					Profile profile = mProfilesById.get(item.getId());
					if (profile != null) {
						playOnDream(profile);
					}
					return kotlin.Unit.INSTANCE;
				}
		);
		load();
	}

	@Override
	public void onDestroy() {
		if (mProgress != null) {
			mProgress.dismiss();
			mProgress = null;
		}
		if (mSimpleResultTask != null)
			mSimpleResultTask.cancel(true);
		super.onDestroy();
	}

	@SuppressWarnings("deprecation")
	private void playOnDream(@NonNull Profile p) {
		String url = null;
		Intent i = getIntent();
		Bundle extras = i.getExtras();
		mShc = SimpleHttpClient.getInstance(p);
		if (Intent.ACTION_SEND.equals(i.getAction()))
			url = extras.getString(Intent.EXTRA_TEXT);
		else if (Intent.ACTION_VIEW.equals(i.getAction()))
			url = i.getDataString();

		if (url != null) {
			Log.i(LOG_TAG, url);
			Log.i(LOG_TAG, p.getHost());

			String time = DateFormat.getDateFormat(this).format(new Date());
			String title = getString(R.string.sent_from_dreamdroid, time);
			if (extras != null) {
				// semperVidLinks sends "artist" and "song" attributes for the
				// youtube video titles
				String song = extras.getString("song");
				if (song != null) {
					String artist = extras.getString("artist");
					if (artist != null)
						title = artist + " - " + song;
				} else {
					String tmp = extras.getString("title");
					if (tmp != null)
						title = tmp;
				}
			}
			mTitle = title;

			Uri uri = Uri.parse(url);
			url = URLEncoder.encode(url).replace("+", "%20");
			title = URLEncoder.encode(title).replace("+", "%20");

			String ref = "4097:0:1:0:0:0:0:0:0:0:" + url + ":" + title;

			if ("youtu.be".equals(uri.getHost())) {
				String vid = uri.getPath().substring(1);
				ref = String.format("8193:0:1:0:0:0:0:0:0:0:%s:%s", URLEncoder.encode(String.format("yt://%s", vid)), title);
			}
			Log.i(LOG_TAG, ref);
			ArrayList<NameValuePair> params = new ArrayList<>();
			params.add(new NameValuePair("file", ref));
			execSimpleResultTask(params);
		} else {
			finish();
		}
	}

	public void load() {
		Profile.ProfileDao dao = AppDatabase.profiles(getContext());
		mProfiles = dao.getProfiles();
		mProfilesById.clear();
		if (mProfiles.size() > 1) {
			ArrayList<ProfileListItem> items = new ArrayList<>();
			for (Profile m : mProfiles) {
				int id = m.getId() == null ? 0 : m.getId();
				mProfilesById.put(id, m);
				items.add(new ProfileListItem(id, m.getName(), m.getHost(), false));
			}
			mListState.replaceAll(items);
		} else {
			if (mProfiles.size() == 1) {
				playOnDream(mProfiles.get(0));
			} else {
				showToast(getString(R.string.no_profile_available));
			}
		}
	}

	@SuppressWarnings("unchecked")
	public void execSimpleResultTask(ArrayList<NameValuePair> params) {
		if (mSimpleResultTask != null) {
			mSimpleResultTask.cancel(true);
		}
		mProgress = ProgressDialog.show(this, getString(R.string.loading), getString(R.string.loading));
		SimpleResultRequestHandler handler = new SimpleResultRequestHandler(URIStore.MEDIA_PLAYER_PLAY);
		mSimpleResultTask = new SimpleResultTask(handler, this);
		mSimpleResultTask.execute(params);
	}

	public void onSimpleResult(boolean success, ExtendedHashMap result) {
		if (mProgress != null) {
			mProgress.dismiss();
			mProgress = null;
		}

		if (mTitle == null)
			mTitle = "...";
		String toastText = getString(R.string.sent_as, mTitle);
		if (mShc.hasError()) {
			toastText = mShc.getErrorText(this);
		}

		showToast(toastText);
		finish();
	}

	public void showToast(String text) {
		Toast toast = Toast.makeText(this, text, Toast.LENGTH_LONG);
		toast.show();
	}

	@NonNull
	@Override
	public Context getContext() {
		return this;
	}
}

File: app/src/net/reichholf/dreamdroid/activities/ShareActivity.kt
/* © 2010 Stephan Reichholf <stephan at reichholf dot net>
 *
 * Licensed under the Create-Commons Attribution-Noncommercial-Share Alike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 */

package net.reichholf.dreamdroid.activities

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.res.stringResource
import net.reichholf.dreamdroid.DreamDroid
import net.reichholf.dreamdroid.R
import net.reichholf.dreamdroid.helpers.LocalNetworkPermissionRequest
import net.reichholf.dreamdroid.ui.share.ShareProfilesHost
import net.reichholf.dreamdroid.ui.share.ShareRequest
import net.reichholf.dreamdroid.ui.share.ShareViewModel
import net.reichholf.dreamdroid.ui.theme.DreamDroidTheme

/**
 * Share / view intent → pick a profile (Compose) → play on the box via MEDIA_PLAYER_PLAY.
 */
class ShareActivity : AppCompatActivity() {
    private val viewModel: ShareViewModel by viewModels()
    private val localNetworkPermissionRequest = LocalNetworkPermissionRequest(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        DreamDroid.setTheme(this)
        enableEdgeToEdge()
        super.onCreate(savedInstanceState)
        localNetworkPermissionRequest.ensure(this)
        title = getText(R.string.watch_on_dream)
        setContent {
            DreamDroidTheme {
                ShareProfilesHost(
                    title = stringResource(R.string.watch_on_dream),
                    state = viewModel.listState,
                    onProfileClick = viewModel::onProfileClick
                )
            }
            val toast = viewModel.toast
            LaunchedEffect(toast) {
                if (toast != null) {
                    Toast.makeText(applicationContext, toast, Toast.LENGTH_LONG).show()
                    viewModel.consumeToast()
                }
            }
            val finished = viewModel.finished
            LaunchedEffect(finished) {
                if (finished) {
                    finish()
                }
            }
        }
        viewModel.start(shareRequest(intent))
    }

    private fun shareRequest(intent: Intent): ShareRequest {
        val extras = intent.extras
        val url = when (intent.action) {
            Intent.ACTION_SEND -> extras?.getString(Intent.EXTRA_TEXT)
            Intent.ACTION_VIEW -> intent.dataString
            else -> null
        }
        // semperVidLinks sends "artist" and "song" for YouTube video titles.
        val song = extras?.getString("song")
        val title = if (song != null) {
            extras.getString("artist")?.let { artist -> "$artist - $song" }
        } else {
            extras?.getString("title")
        }
        return ShareRequest(url = url, title = title)
    }
}
